compiler/src/iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp

```diff
@@ -835,27 +835,71 @@ OpFoldResult TensorReshapeOp::fold(FoldAdaptor operands) {
+
+OpFoldResult TensorBitCastOp::fold(FoldAdaptor operands) {
+  auto sourceType = llvm::cast<ShapedType>(getSource().getType());
+  auto resultType = llvm::cast<ShapedType>(getResult().getType());
+  if (sourceType.getElementType() != resultType.getElementType()) {
+    // Element type mismatch, this is a bitcast.
+    return {};
+  }
+  if (compareShapesEqual(sourceType, getSourceDims(), resultType,
+                         getResultDims())) {
+    // Shapes match and this is a no-op so just fold to the source.
+    return getSource();
+  }
+  return {};
+}
+
 namespace {
 
-// Flatten a chain of reshapes (reshape feeding into reshape) such that a
-// reshape only ever pulls from a non-reshape source. This prevents big useless
-// chains and makes it easier to track the original storage for the tensor.
-struct FlattenTensorReshapeChain : public OpRewritePattern<TensorReshapeOp> {
-  using OpRewritePattern<TensorReshapeOp>::OpRewritePattern;
+// Flatten a chain of reshapes or bitcasts (reshape/bitcast feeding into
+// reshape or bitcast) such that a reshape only ever pulls from a non-reshape
+// source. This prevents big useless chains and makes it easier to track the
+// original storage for the tensor.
+template <typename CastOpTy>
+struct FlattenTensorCastLikeChain : public OpRewritePattern<CastOpTy> {
+  using OpRewritePattern<CastOpTy>::OpRewritePattern;
 
-  LogicalResult matchAndRewrite(TensorReshapeOp reshapeOp,
+  LogicalResult matchAndRewrite(CastOpTy reshapeOp,
                                 PatternRewriter &rewriter) const override {
-    auto sourceOp = dyn_cast_or_null<TensorReshapeOp>(
-        reshapeOp.getSource().getDefiningOp());
-    if (!sourceOp)
-      return failure();
-
     // We want the same result value/shape but to source from the ancestor. We
     // need to pull any dynamic dims from that as we don't care about the
     // intermediate reshapes.
-    rewriter.replaceOpWithNewOp<TensorReshapeOp>(
-        reshapeOp, reshapeOp.getResult().getType(), sourceOp.getSource(),
-        sourceOp.getSourceDims(), reshapeOp.getResultDims());
+    Value source;
+    ValueRange sourceDims;
+    if (auto sourceOp = dyn_cast_or_null<TensorReshapeOp>(
+            reshapeOp.getSource().getDefiningOp())) {
+      source = sourceOp.getSource();
+      sourceDims = sourceOp.getSourceDims();
+    } else if (auto sourceOp = dyn_cast_or_null<TensorBitCastOp>(
+                   reshapeOp.getSource().getDefiningOp())) {
+      source = sourceOp.getSource();
+      sourceDims = sourceOp.getSourceDims();
+    }
+
+    if (!source) {
+      return failure();
+    }
+
+    auto sourceType = llvm::cast<ShapedType>(source.getType());
+    auto resultType = llvm::cast<ShapedType>(reshapeOp.getResult().getType());
+
+    // If the bitwidths don't match, this is a bitcast, else we can use
+    // reshape.
+    if (IREE::Util::getTypeBitWidth(sourceType.getElementType()) !=
```

Comment:
do we want to just say if element types differ it's a bitcast? (that way when element types are structs and other things stuff works - as-is if we had tensor<4xi8> and tensor<1x<my_struct_of_32_bits>> this would still be a reshape and that doesn't seem right)

Reply:
`tensor<4xi8> and tensor<1x<my_struct_of_32_bits>>` would be a bitcast (because 8 -> 32) but, yeah, `i32` -> `struct_of_32_bits` does sound weird. Also it's a little confusing at the moment that reshape is allowed to bitcast as well. Sounds like a good change, should I disallow bitcasts on reshapes?

Reply:
(oh unless you meant element type of `tensor<4xi8>`, which sounds interesting...)